Tools/Algorithms/psatd.ipynb

```diff
@@ -0,0 +1,631 @@
+    "- `rho_in_time` (str: `'constant'`, `'linear'`, `'quadratic'`): $\\rho$ constant, linear, or quadratic in time.\n",
+    "\n",
+    "In the notebook, the constant, linear, and quadratic coefficients of $\\boldsymbol{J}$ and $\\rho$ are denoted with the suffixes `_c0`, `_c1`, `_c2`, respectively. However, the corresponding coefficients in the displayed equations are denoted with the prefixes $\\gamma$, $\\beta$, and $\\alpha$, respectively. For example, if $\\rho$ is quadratic in time, it will be denoted as `rho = rho_c0 + rho_c1*(t-tn) + rho_c2*(t-tn)**2` in the notebook and $\\rho(t) = \\gamma_{\\rho} + \\beta_{\\rho} (t-t_n) + \\alpha_{\\rho} (t-t_n)^2$ in the displayed equations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "divE_cleaning = True\n",
+    "divB_cleaning = True\n",
+    "J_in_time = 'constant'\n",
+    "rho_in_time = 'constant'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Auxiliary functions:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def check_diag(W, D, P, invP):\n",
+    "    \"\"\"\n",
+    "    Check diagonalization of W as P*D*P**(-1).\n",
+    "    \"\"\"\n",
+    "    Wd = P * D * invP\n",
+    "    for i in range(Wd.shape[0]):\n",
+    "        for j in range(Wd.shape[1]):\n",
+    "            Wd[i,j] = Wd[i,j].expand().simplify()\n",
+    "            diff = W[i,j] - Wd[i,j]\n",
+    "            diff = diff.expand().simplify()\n",
+    "            assert (diff == 0), f'Diagonalization failed: W[{i},{j}] - Wd[{i},{j}] = {diff} is not zero'\n",
+    "\n",
+    "def simple_mat(W):\n",
+    "    \"\"\"\n",
+    "    Simplify matrix W.\n",
+    "    \"\"\"\n",
+    "    for i in range(W.shape[0]):\n",
+    "        for j in range(W.shape[1]):\n",
+    "            W[i,j] = W[i,j].expand().simplify()\n",
+    "            \n",
+    "def limit(f, x, l):\n",
+    "    \"\"\"\n",
+    "    Compute limit of f at x=l.\n",
+    "    \"\"\"\n",
+    "    lim = f.series(x, l, 1).removeO() - f.series(x, l, 0).removeO()\n",
+    "    lim = lim.simplify()\n",
+    "    return lim"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Definition of symbols:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define dimensionality parameter used throughout the notebook\n",
+    "dim = 6\n",
+    "if divE_cleaning:\n",
+    "    dim += 1\n",
+    "if divB_cleaning:\n",
+    "    dim += 1\n",
+    "\n",
+    "# Define symbols for physical constants\n",
+    "c = sp.symbols(r'c', real=True, positive=True)\n",
+    "mu0 = sp.symbols(r'\\mu_0', real=True, positive=True)\n",
+    "\n",
+    "# Define symbols for time variables\n",
+    "s = sp.symbols(r's', real=True, positive=True)\n",
```

Comment:
Could you add a comment explaining what `s` is here?

Reply:
Thank you. Comment added in ff98459, let me know if it isn't clear enough.